File: 3-1/nypc/로봇청소기.py
import sys
input=sys.stdin.readline

v=[]

def sorting():
    tsize=len(v)
    for i in range(tsize):
        for j in range(i+1, tsize):
            if v[i][0][0] > v[j][0][0]:
                v[i], v[j] = v[j], v[i]
n, m = map(int,input().split())
map=[[]]
for i in range(n):
    map.append(list(" "+input()))

for i in range(n, 0, -1):
    sorting()
    print(v)
    for vi in range(len(v)):
        posx=v[vi][0][0]
        deltapos=v[vi][0][1]-i

        for j in range(max(posx-deltapos, 1), min(posx+deltapos, m)+1):
            if map[i][j]=='X':
                map[i][j]='O'
                for k in range(deltapos-abs(posx-j)):
                    v[vi][1].append('D')
                if (posx-j>0):
                    for k in range(posx-j):
                        v[vi][1].append('R')
                else:
                    for k in range(j-posx):
                        v[vi][1].append('L')

                v[vi][0][0]=j
                v[vi][0][1]=i
                break
    for j in range(1, m+1):
        if map[i][j]=='X':
            map[i][j]='O'
            v.append([[j, i], ['D']*(n-i)])
for i in range(len(v)):
    for j in range(0, n-1-len(v[i])+1):
        v[i][1].append('D')

print(len(v))
for i in range(len(v)):
    print(v[i][0][0], end=' ')
    for _ in range(n-2, -1, -1):
        print(v[i][1][_], end='')
    print()
      

File: 3-1/nypc/로봇청소기.cpp
#include <stdio.h>
#include <string.h>
#include <algorithm>
#include <stdlib.h>
#include <queue>
#define abs(a) ((a)>0?(a):-(a))
using namespace std;
typedef pair<int,int> pii;
typedef long long int lli;

vector<pair<pii, vector<char> > > v;
int n, m;
char map[100000][100000]; 

void sort(){
	int tsize=v.size();
	for(int i=0;i<tsize;i++){
		for(int j=i+1;j<tsize;j++){
			if(v[i].first.first>v[j].first.first){
				swap(v[i], v[j]);
			}
		}
	}
}

int main(){
	scanf("%d %d", &n, &m);
	for(int i=1;i<=n;i++) scanf("%s", &map[i][1]);
	
	for(int i=n;i>0;i--){
//		printf("error");
//		sort();
		for(int vi=0;vi<v.size();vi++){
			int posx=v[vi].first.first, deltapos=v[vi].first.second-i;
			int mxp=posx+deltapos<m?posx+deltapos:m;
			for(int j=posx-deltapos>0?posx-deltapos:1; j<=mxp; j++){
				if(map[i][j]=='X'){
					map[i][j]='O';
					for(int k=0;k<deltapos-abs(posx-j);k++){
						v[vi].second.push_back('D');
					}
					if(posx-j>0){
						for(int k=0;k<posx-j;k++) v[vi].second.push_back('R');
					}
					else{
						for(int k=0;k<j-posx;k++) v[vi].second.push_back('L');
					}
					
					v[vi].first.first=j;
					v[vi].first.second=i;
					break;
				}
			}
		}
		for(int j=1;j<=m;j++){
			if(map[i][j]=='X'){
				map[i][j]='O';
				vector<char> tempv;
				for(int k=0;k<n-i;k++){
					tempv.push_back('D');
				}
				v.push_back(make_pair(make_pair(j, i), tempv));
			}
		}	
	}
	
	for(int i=0;i<v.size();i++){
		while(v[i].second.size()<n-1) v[i].second.push_back('D');
	}
	
	printf("%d\n", v.size());
	for(int i=0;i<v.size();i++){
		printf("%d ", v[i].first.first);
		for(int j=n-2;j>=0;j--) printf("%c", v[i].second[j]);
		printf("\n");
	}
	
	return 0;
}
